sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JDBCRDD.scala

```diff
@@ -301,10 +306,44 @@ private[jdbc] class JDBCRDD(
         ResultSet.TYPE_FORWARD_ONLY, ResultSet.CONCUR_READ_ONLY)
     stmt.setFetchSize(options.fetchSize)
     stmt.setQueryTimeout(options.queryTimeout)
-    rs = stmt.executeQuery()
+
+    val isFinished = new AtomicBoolean(false)
+    startTheJobInterruptListenerThread(context, stmt, isFinished)
+    try{
+      rs = stmt.executeQuery()
+    } finally {
+      isFinished.set(true)
+    }
     val rowsIterator = JdbcUtils.resultSetToSparkInternalRows(rs, schema, inputMetrics)
 
     CompletionIterator[InternalRow, Iterator[InternalRow]](
       new InterruptibleIterator(context, rowsIterator), close())
   }
+
+  /**
+   * Start the job interruption listener thread, if the job is interrupted and
+   * stmt.executeQuery() is not completed, initiate a stmt.cancel() request.
+   */
+  private def startTheJobInterruptListenerThread(context: TaskContext, stmt: PreparedStatement
+                                                 , isFinished: AtomicBoolean): Unit = {
```

Comment:
I think we already respect the interruption via `InterruptibleIterator`:
https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JDBCRDD.scala#L308

Why do we need to handle it in this way?

Reply:
InterruptibleIterator is only for sql execution completed and returned ResultSet object, support interruption in the process of data iteration. When sql is executing, it cannot be interrupted

Reply:
How about setting `fetchSize`? If ResultSet is a cursor, it seems the process can be interrupted correctly.

Reply:
Okay, I will test and verify it at night. But I think it might not work. In view of the complicated logic and the execution of SQL that takes several minutes or even hours, the ResultSet object may not be returned in time.

Reply:
I found that by setting FetchSize, can respond quickly and return the ResultSet object. If it is mysql database, you need to add the useCursorFetch=true parameter in the jdbc url. Thank you for your answer, I will close this pull request.

Reply:
I'm sorry I opened this question again. Because there were a lot of things in the previous two days, I closed this question in a hurry.
If a more complex query sql specifies the fetchsize parameter, it is also invalid, because sql logic processing takes a long time, so the ResultSet cursor will not be returned immediately.

Reply:
Please describe what's an issue in the PR description, first.

Reply:
Already added, thank you!

Reply:
This pull request is useful for long-running services that often deal with jdbc data sources.

Reply:
I'm still not sure whether this is useful or not. If `fetchSize` is not enough, how about using `queryTimeout` together?

Reply:
In real scenarios, `query` is often a more complicated SQL. This SQL may use joins between several tables or views. Setting `fetchsize` can not return the cursor object immediately, setting `queryTimeout` cannot satisfy , The client will stop the request immediately after it stops, and the job will be abnormally terminated due to the unreasonable setting of `queryTimeout`.

Reply:
Try to write a complex SQL, such as Cartesian product to test the effect of jdbc `fetchsize`，Or based on the jdbc query of hive on mapreduce, you will find that the `fetchsize` is set at this time, and you have to wait for the job to execute for a period of time before returning the cursor object. The client cannot terminate the job run.